### DISCUSSION TITLE
Waveshare ST7789 - Read function not working & Animated Dial not working either

### DISCUSSION THREAD

**Discussion by phillmj** [Bodmer/TFT_eSPI#2237] [Q&A - General] [answered]
1. This display supports read operation (according to lengthy technical manual, linked below), but when I try the Read/WriteTest example it gives errors. Also the Sprite Animated Dial does not work, with lots of other errors but I'm assuming one of the main ones is the read function isn't working. The SPI read frequency is set to 20000000

The screen: https://www.waveshare.com/wiki/2inch_LCD_Module#Feature

The technical manual for this screen is here: https://www.waveshare.com/w/upload/a/ad/ST7789VW.pdf

3. IDE: Arduino Teensy
4. TFT_eSPI library version: 2.4.79
5. Procesor:Teensy 4.0
6. TFT driver: ST7789
7. Interface type: SPI

Unfortunately I do not understand any of this well enough to edit the library to solve it. Is there a way to fix this? I have spent a long time googling so apologies if this has been answered.

Error from the ReadWriteDiagnostic:

```
 Pixel value written = 1
 Pixel value read    = 0
 ERROR                 ^^^^
 Pixel value written = 2
 Pixel value read    = 0
 ERROR                 ^^^^
 Pixel value written = 4
 Pixel value read    = 0
 ERROR                 ^^^^
 Pixel value written = 8
 Pixel value read    = 0
 ERROR                 ^^^^
 Pixel value written = 10
```

Arduino/TEensy errors from trying to run the Animated Dial:
```
Arduino: 1.8.19 (Mac OS X), TD: 1.56, Board: "Teensy 4.0, MIDI, 600 MHz, Fast, US English"

In file included from /Users/Mikey/Documents/Arduino/libraries/TFT_eSPI/TFT_eSPI.cpp:5556:0:
/Users/Mikey/Documents/Arduino/libraries/TFT_eSPI/Extensions/Sprite.cpp: In member function 'void TFT_eSprite::scroll(int16_t, int16_t)':
/Users/Mikey/Documents/Arduino/libraries/TFT_eSPI/Extensions/Sprite.cpp:1425:15: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
   if (abs(dx) >= _sw || abs(dy) >= _sh)
               ^
/Users/Mikey/Documents/Arduino/libraries/TFT_eSPI/Extensions/Sprite.cpp:1425:33: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
   if (abs(dx) >= _sw || abs(dy) >= _sh)
                                 ^
/Users/Mikey/Documents/Arduino/libraries/TJpg_Decoder/src/TJpg_Decoder.cpp: In member function 'JRESULT TJpg_Decoder::drawSdJpg(int32_t, int32_t, const String&)':
/Users/Mikey/Documents/Arduino/libraries/TJpg_Decoder/src/TJpg_Decoder.cpp:397:28: error: no matching function for call to 'SDClass::exists(const String&)'
   if ( !SD.exists(pFilename) )
                            ^
In file included from /Users/Mikey/Documents/Arduino/libraries/TJpg_Decoder/src/TJpg_Decoder.h:36:0,
                 from /Users/Mikey/Documents/Arduino/libraries/TJpg_Decoder/src/TJpg_Decoder.cpp:10:
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/SD/src/SD.h:189:7: note: candidate: virtual bool SDClass::exists(const char*)
  bool exists(const char *filepath) {
       ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/SD/src/SD.h:189:7: note:   no known conversion for argument 1 from 'const String' to 'const char*'
/Users/Mikey/Documents/Arduino/libraries/TJpg_Decoder/src/TJpg_Decoder.cpp:403:57: error: no matching function for call to 'SDClass::open(const String&, int)'
     return drawSdJpg(x, y, SD.open( pFilename, FILE_READ));
                                                         ^
In file included from /Users/Mikey/Documents/Arduino/libraries/TJpg_Decoder/src/TJpg_Decoder.h:36:0,
                 from /Users/Mikey/Documents/Arduino/libraries/TJpg_Decoder/src/TJpg_Decoder.cpp:10:
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/SD/src/SD.h:181:7: note: candidate: virtual File SDClass::open(const char*, uint8_t)
  File open(const char *filepath, uint8_t mode = FILE_READ) {
       ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/SD/src/SD.h:181:7: note:   no known conversion for argument 1 from 'const String' to 'const char*'
/Users/Mikey/Documents/Arduino/libraries/TJpg_Decoder/src/TJpg_Decoder.cpp: In member function 'JRESULT TJpg_Decoder::getSdJpgSize(uint16_t*, uint16_t*, const String&)':
/Users/Mikey/Documents/Arduino/libraries/TJpg_Decoder/src/TJpg_Decoder.cpp:460:28: error: no matching function for call to 'SDClass::exists(const String&)'
   if ( !SD.exists(pFilename) )
                            ^
In file included from /Users/Mikey/Documents/Arduino/libraries/TJpg_Decoder/src/TJpg_Decoder.h:36:0,
                 from /Users/Mikey/Documents/Arduino/libraries/TJpg_Decoder/src/TJpg_Decoder.cpp:10:
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/SD/src/SD.h:189:7: note: candidate: virtual bool SDClass::exists(const char*)
  bool exists(const char *filepath) {
       ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/SD/src/SD.h:189:7: note:   no known conversion for argument 1 from 'const String' to 'const char*'
/Users/Mikey/Documents/Arduino/libraries/TJpg_Decoder/src/TJpg_Decoder.cpp:466:60: error: no matching function for call to 'SDClass::open(const String&, int)'
     return getSdJpgSize(w, h, SD.open( pFilename, FILE_READ));
                                                            ^
In file included from /Users/Mikey/Documents/Arduino/libraries/TJpg_Decoder/src/TJpg_Decoder.h:36:0,
                 from /Users/Mikey/Documents/Arduino/libraries/TJpg_Decoder/src/TJpg_Decoder.cpp:10:
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/SD/src/SD.h:181:7: note: candidate: virtual File SDClass::open(const char*, uint8_t)
  File open(const char *filepath, uint8_t mode = FILE_READ) {
       ^
/Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/SD/src/SD.h:181:7: note:   no known conversion for argument 1 from 'const String' to 'const char*'
Multiple libraries were found for "TJpg_Decoder.h"
 Used: /Users/Mikey/Documents/Arduino/libraries/TJpg_Decoder
 Not used: /Users/Mikey/Documents/Arduino/libraries/TJpg_Decoder-master
Multiple libraries were found for "SD.h"
 Used: /Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/SD
 Not used: /Applications/Teensyduino.app/Contents/Java/libraries/SD
Error compiling for board Teensy 4.0.


This report would have more information with
"Show verbose output during compilation"
option enabled in File -> Preferences.
```

**Comment by Bodmer** [comment]
There are a few reasons why it does not work.
1. The circuit board has a DIN pin and no DO (data out = MISO).
2. Though the ST7789 controller supports half duplex reads (typically the data pin is then called SDA), the IM pins of the controller may not be wired for this mode on the PCB.
3. The library has no optimisation or specific feature support for half duplex (SDA) mode for the Teensy, and thus uses a bit bashing approach. Since the Teensy is a very fast processor and the ST7789 requires a low read speed (circa 6MHz) it is likely to run too fast.
To progress on this, find out if that display board is configured by the supplier to permit reads.

**Comment by Bodmer** [comment, marked as answer]
I see this in the suppliers web page:

"Note: Different from the traditional SPI protocol, the data line from the slave to the master is hidden since the device only has display requirement."

So that suggests the board does not support reads.

**Reply by phillmj** [reply]
Thank-you so much for your valuable time and expertise. It's a shame, because it's a perfect screen for the project I'm doing, and waveshare is such a large display company.

I will reach out to them directly and inquire about this. You've been very helpful thanks.

**Comment by phillmj** [comment]
I have reached out to Waveshare with this issue.

In the manual https://www.waveshare.com/w/upload/a/ad/ST7789VW.pdf on page 58 there is discussion of read capability. Does this help at all?

Thanks,

Mike

**Reply by Bodmer** [reply]
No, it is how it is wired and the settings of the IMx config pin. Ask for a schematic

**Reply by phillmj** [reply]
Thank-you. Here are two additional documents, but even though one is called the schematic, it's minimal.

- https://www.waveshare.com/w/upload/e/ee/2inch_LCD_Module_SchDoc.pdf
- https://www.waveshare.net/w/upload/b/b1/2inch_LCD_Module.pdf

I will update this thread when I hear back from Waveshare.

Thank-you so much for your time,

Mike

![Screen Shot 2022-12-07 at 9 10 27 AM](https://user-images.githubusercontent.com/97892640/206245180-7eb3c114-53bf-43ed-995b-7bfefee5d6e7.jpg)

**Reply by phillmj** [reply]
I heard back from Waveshare.

Sadly you are right; this screen does not support read operations.

If you have any suggestions for 2.0 or 2.2" 320x240 screens that will work with your library and support reads, I'm all ears.

Thanks again for your time.

What an utterly fantastic library you have created. You should be very proud.

Many thanks,

Mike

**Reply by Bodmer** [reply]
You should be able to manage without needing to read from the TFT.

For example here is an [animated dial sketch](https://github.com/Bodmer/TFT_eSPI/discussions/860#discussioncomment-208303) (myDZM_array3) that uses sprites to maintain a copy of the screen areas that need to be updated and refreshed. No reads from the screen RAM are used and it runs faster than reading the screen.

**Comment by phillmj** [comment]
That is excellent, thanks so much!

Question - is it possible to push a sprite into another sprite? I notice there is a pushToSprite option but didn't see any documentation. It would be so cool to basically "package" several sprites into one larger one for example.

Also, last q, are smaller sprites faster to render than larger sprites? I'm having some performance issues with sprites but I used the entire screen because it was easier to position things.

Thanks again,

Mike

**Reply by Bodmer** [reply]
```
// Push the sprite to another sprite at x,y. This fn calls pushImage() in the destination sprite (dspr) class.
  bool     pushToSprite(TFT_eSprite *dspr, int32_t x, int32_t y);
  bool     pushToSprite(TFT_eSprite *dspr, int32_t x, int32_t y, uint16_t transparent);
```
Example use:
```
// Push sprite spr1 to spr2 at position 5, 10
spr1.pushToSprite(&spr2, 5, 10);

// As above but all blue pixels in spr1 will not be rendered in spr2, so are effectively transparent (can be any colour of choice)
spr1.pushToSprite(&spr2, 5, 10, TFT_BLUE);
```

**Reply by phillmj** [reply]
That is SO helpful! Thank you so much!

**Reply by phillmj** [reply]
By the way - this trick alone let me do some pretty amazing things with your library! So thank-you again!

Also, a question - is there a way to pushToSpriteRotated? I want to be able to push a rotated sprite into another sprite...

Thanks again!!!

Mike

**Reply by Bodmer** [reply]
Yes.
```
          // Push a rotated copy of Sprite to another different Sprite with optional transparent colour
  bool     pushRotated(TFT_eSprite *spr, int16_t angle, uint32_t transp = 0x00FFFFFF);
```
Make sure you set the pivot points for the source sprite and the destination sprite.

**Reply by phillmj** [reply]
Mind-blowing! "WOW!!!" Doesn't begin to cover it! Thank-you so much!!

Thanks to you, AGAIN!, I've been able to do a really cool thing for the project I'm doing.

I noticed that rotated sprites sometimes render graphics somewhat pixelated - for example, text that is clear when not rotated has jaggies. Maybe this is due to the screen resolution (I'm using 320 x 240), and not an issue with the library.

Here's a screenshot - you'll see the pixelation on the rotated numbers - but I'm not attaching for that reason. I'm attaching it to say, wow, thanks, I never could have done this without your genius!

All the best,

Mike
![IMG_1873](https://user-images.githubusercontent.com/97892640/208222060-64773155-41bb-425e-9b89-52e35416e594.jpeg)



All in all - fantastic, thanks so much Mr. Bodmer!!

Mike

**Reply by Bodmer** [reply]
Great, looks very good, thanks for sharing.

The rotation code does not perform bilinear interpolation at the moment, so does not produce anti aliased results. I rarely use the function so have not got around to upgrading it. It is on my long to-do list!

**Comment by phillmj** [comment]
I completely understand.

Many thanks for your time. I hope you are truly well paid for your genius.
